### ISSUE TITLE
Connect lightgbm learner

### ISSUE THREAD

**Issue by kapsner** [mlr-org/mlr3learners#57]
Hello,

I have experimented a little during the last weeks and maybe this is of interest for someone:

I was able to implement the [lightgbm python module](https://github.com/microsoft/LightGBM/tree/master/python-package) in the R package [lightgbm.py](https://github.com/kapsner/lightgbm.py) and again use this "base-implementation" as dependency for the R package [mlr3learners.lightgbm](https://github.com/kapsner/mlr3learners.lightgbm).

It works so far and I have added vignettes for binary classification and multiclass classification examples as well as a regression task.

The base-implementation uses [reticulate](https://github.com/rstudio/reticulate) as an R interface to the python module.

The mlr3-extension currently not passes the `run_autotests` test, so some debugging need so be done. However, if this approach is of interest for someone, it could be a possibility of bringing the great LightGBM to the mlr3 framework.

**Comment by pat-s** [comment]
@kapsner Thanks for contributing! 

We always welcome new learners :) 

I've just added some more detailed instructions to the [README](https://github.com/mlr-org/mlr3learners#requestingadding-additional-learners) that should guide you through the process of adding a new learner to the org.
We would like to collect all mlr3learners in an org so that users can easily browse them.

In your case, you would be at point 3.

Feel free to ask any question if something might be unclear.

**Comment by kapsner** [comment]
@pat-s Thanks for adding more detailed instructions to the README.

I am still working on some minor fixes and once done, I am going to transfer the project to the mlr3learners org.

I order to avoid misconception with the ["native" lightgbm R package](https://github.com/microsoft/LightGBM/tree/master/R-package) (which is unfortunatelly not on CRAN yet), I have renamed my repo to [mlr3learners.lgbpy](https://github.com/kapsner/mlr3learners.lgbpy) to somehow indicate already in the package name, that it is using lightgbm's python module.

According to [this note](https://github.com/microsoft/LightGBM/issues/2640#issuecomment-567508599), using the python module via `reticulate` could come with some performance drawbacks and might not be the preferred way and therefore always a workaround until an easy-to-install lightgbm R package exists.

**Comment by pat-s** [comment]
Thanks. Renaming makes sense.

Have you done any benchmarks yet between your version and the _lightgbm_ R package? I wonder if its worth the effort continuing connecting the python package if there is already effort to have a native R implementation?
This one could also then easily be wrapped into a mlr3-learner.

I do not have an overview of what's going on for this learner in detail so just let me know how you want to proceed. We do not mind having multiple learners interfacing the same algorithm since it is up to the user to decided which one to use.

**Comment by kapsner** [comment]
No, I haven't done any benchmarkings yet.

I think, the python approach could be seen as some kind of 'proof-of-concept'. In my opinion, the effort depends on the goal:
- Using reticulate and python should result in an R package, which is easy to install/configure on different operating systems.
- Using the native implementation requires manual installation of prerequisites.

Unfortunatelly, one cannot even install lightgbm with
```r
devtools::install_github("microsoft/LightGBM", subdir = "R-package")
```
but has to go through https://github.com/microsoft/LightGBM/tree/master/R-package.

I am not sure, if the mlr3learners should all be "standalone" R-packages which automatically resolve dependencies?

For now, I have started to also create a mlr3-learner for the native implementation: [mlr3learners.lightgbm](https://github.com/kapsner/mlr3learners.lightgbm)

**Comment by pat-s** [comment]
> I am not sure, if the mlr3learners should all be "standalone" R-packages which automatically resolve dependencies?

When installing the learner package, the learner should be usable afterwards. So yes, the mlr3learners are small, standalone packages that also describe their dependencies.

> For now, I have started to also create a mlr3-learner for the native implementation: mlr3learners.lightgbm

Thanks for the effort! Let us know when you're done.

**Comment by kapsner** [comment]
Hi @pat-s ,

I just tried to transfer both packages to the [mlr3learners organization](https://github.com/mlr3learners) but it said, that I do not have rights to create public repos there, so I am not sure, if it worked.

Here a summary of what is implemented in both R packages:
* Both packages are implementations of the [lightgbm algorithm](https://github.com/microsoft/LightGBM).
* Both packages can make use of a previously GPU-compiled version of lightgbm (a short installation instruction and vignette is available with both packages, tested on a NVIDIA(R) RTX 2060).

In both packages, the algorithm is implemented in the following manner:
1. by default, a 5-fold cross-validation (CV) is performed with the given parameters to find the best number of boosting iterations (`'nrounds'`-argument of the
R-package implementation and `'num_boost_rounds'` of the pyhton implementation)
2. in a second step, a final model is trained with this number of boosting iterations
3. the folds of step 1 can be adjusted
4. the upstream CV can be switched off to only use the model training with a user-defined number of boosting iterations
5. by default, the native multi-threading of the algorithms is used. However, the multi-threading can be configured using the parameter `'num_threads'` and even switched off by setting `'num_threads' = 1`

The advantage of the upstream CV approach is that the argument [`'early_stopping_round'`](https://lightgbm.readthedocs.io/en/latest/Parameters.html#early_stopping_round) can be used to find the optimal number of boosting iterations.
Using this argument, the training is stopped if the performance metric does not improve over a given number of rounds.

* Both packages integrate well with mlr3's parallelization approach by using the `future` R package, when the parameter `'num_threads'` is set to 1. An example of using the future parallelization exists in both packages in the "tuning_example" vignette. 

* Furthermore, there are vignettes for binary classification, multiclass classification and regression tasks in each package.

The package [`mlr3learners.lightgbm`](https://github.com/kapsner/mlr3learners.lightgbm) is a mlr3 wrapper around the [lightgbm R-package](https://github.com/microsoft/LightGBM/tree/master/R-package). This original R implementation is unfortunatelly not fully self-contained,
since the lightgbm R package is neither on CRAN nor automatically installable via `devtools`.

Therefore I created the package [`mlr3learners.lgbpy`](https://github.com/kapsner/mlr3learners.lgbpy), which basically provides an mlr3 interface to the [lightgbm python-package](https://github.com/microsoft/LightGBM/tree/master/python-package via the [`reticulate`](https://rstudio.github.io/reticulate/) R package. Hence it is installable 'standalone' without the need to prior compile the lightgbm learner. Instead, the installation works via a python/(mini-)conda environment, similar to packages like e.g. [`tensorflow`](https://tensorflow.rstudio.com/installation/) or [`h2o4gpu`](https://www.h2o.ai/blog/h2o4gpu-now-available-in-r/).

* In both packages, the mlr3 tests are working.

**Comment by pat-s** [comment]
Thanks for the detailed update.
I invited you to the org. You can now transfer the repo and then I'll give you full access to the repo.

Regarding "early stopping": We want to implement a general approach for this in {mlr3tuning}: https://github.com/mlr-org/mlr3tuning/issues/216

I am not sure if we should add both packages. Which one do you think will get longterm support by yourself? Do you really want to maintain two highly redundant packages in the longterm? Just trying to save you from too much burden here :)

I think both approaches have their pros/con: 

- for light.gbm you devote more work to the user but remove the intermediate python package layer
- for lgbpy you make installation more easy but also insert more deps that can make you unflexible and could potentially cause bugs

On a side note, are you aware that you can list python deps directly in the DESC file since [reticulate v1.1.4](https://github.com/rstudio/reticulate/blob/master/NEWS.md#reticulate-114)?

You are welcome to transfer both packages to the mlr3learners org but then we'd also like that you maintain both in a working state :)

Whatever you decide for, please edit the mlr3learners wiki after having transferred the repo(s) with the packages and list them in the appropriate section. Thanks!

**Comment by kapsner** [comment]
> Regarding "early stopping": We want to implement a general approach for this in {mlr3tuning}: mlr-org/mlr3tuning#216

That's good to hear. I already thought of adapting the early-stopping approach I implemented in the lightgbm-mlr3-learners to xgboost, which would then be superfluous.
And the mlr3-tuning-approach should not interfere with the internal CV-approach for early-stopping, because it is already moved to an extra method, which could easily be removed from the workflow.

> you can list python deps directly in the DESC file since reticulate v1.1.4?
Already using this feature in the dependency-package of `mlr3learners.lgbpy`: https://github.com/kapsner/lightgbm.py/blob/master/DESCRIPTION :)

> Which one do you think will get longterm support by yourself?
Since the `mlr3learners.lightgbm` uses the native R implementation, this should be the package with the longterm support. However, to be a selfcontained R package, it depends on the progess of https://github.com/microsoft/LightGBM/issues/1910 and https://github.com/microsoft/LightGBM/issues/629.
For now, I have decided to transfer only the `mlr3learners.lightgbm` package. However, the other one could be easily added as well, if required.

**Comment by t-wojciech** [comment]
LightGBM is now on CRAN: https://cran.r-project.org/web/packages/lightgbm/index.html

**Comment by berndbischl** [comment]
@t-wojciech great! we will try to integrate as soon as possible!

**Comment by kapsner** [comment]
https://github.com/mlr3learners/mlr3learners.lightgbm is already using the CRAN version: https://github.com/mlr3learners/mlr3learners.lightgbm/blob/master/DESCRIPTION#L30

The package has been there already this year but was again removed due to failing CRAN tests: https://github.com/microsoft/LightGBM/issues/629#issuecomment-705331560

**Comment by t-wojciech** [comment]
Good to know! I [added](https://github.com/microsoft/LightGBM/pull/3666) a `mlr3learners.lightgbm` to `lightgbm` README, so it can reach more users.
